examples/gno.land/r/demo/gnofundme/gnofundme.gno

```diff
@@ -0,0 +1,253 @@
+	users          *avl.Tree
+	campaignPrefix = "/r/demo/gnofundme:"
+)
+
+func init() {
+	campaigns = avl.NewTree()
+	users = avl.NewTree()
+}
+
+func NewUser(name string, email string) string {
+	// get the user address
+	address := std.GetOrigCaller()
+	// create the user
+	user := pkg.NewUser(name, email, address)
+
+	// the avl.tree key will be address to simplify the fetching
+	users.Set(address.String(), user)
+
+	return user.GetName() + " added successfully"
+}
+
+func NewCampaign(title string, description string, goal uint, beginT int64, deadlineT int64, beneficiary std.Address) (int, string) {
+	// get the caller as the owner address
+	caller := std.GetOrigCaller()
+	user := pkg.NewUser("user1", "email1", caller)
+
+	// the avl.tree key will be address to simplify the fetching
+	users.Set(caller.String(), user)
+
+	user = pkg.NewUser("user2", "email2", beneficiary)
+
+	// the avl.tree key will be address to simplify the fetching
+	users.Set(beneficiary.String(), user)
+	// verify that the user exists
+	o, ok := users.Get(caller.String())
+	if !ok {
+		panic("owner not found")
+	}
+
+	// verify that the benefeciary exists
+	b, ok := users.Get(beneficiary.String())
+	if !ok {
+		panic("beneficiary not found")
+	}
+
+	// cast the owner and beneficiary to *pkg.User
+	owner := o.(*pkg.User)
+	beneficiaryU := b.(*pkg.User)
+
+	// create the time objects
+	begin := time.Unix(beginT, 0)
+	deadline := time.Unix(deadlineT, 0)
+
+	// create the campaign
+	campaign := pkg.NewCampaign(title, description, goal, begin, deadline, owner, beneficiaryU)
+
+	// the key will be the size of the tree
+	key := campaigns.Size()
+	campaigns.Set(strconv.Itoa(key), campaign)
+
+	return key, "campaign created successfully"
+}
+
+func AddContributor(campaignId string, amount uint, display bool) string {
+	// get the campaign if it exists
+	campaign, ok := campaigns.Get(campaignId)
+	if !ok {
+		return "campaign not found"
+	}
+
+	// get the user address
+	userAddress := std.GetOrigCaller()
+
+	// get the user if he exists
+	user, ok := users.Get(userAddress.String())
+	if !ok {
+		panic("user is not registered")
+	}
+
+	sentCoins := std.GetOrigSend()
+	// amount := sentCoins.AmountOf("ugnot")
+
+	// add the contributor to the campaign
+	campaign.(*pkg.Campaign).AddContributor(user.(*pkg.User), amount, display)
+
+	return "contributor added successfully"
+}
+
+func GetCampaigns() string {
+	if campaigns.Size() == 0 {
+		return "[]"
+	}
+	campaignsArray := make([]*pkg.Campaign, 0, campaigns.Size())
+	campaigns.Iterate("", "", func(key string, value interface{}) bool {
+		c := value.(*pkg.Campaign)
+		campaignsArray = append(campaignsArray, c)
+		return false
+	})
+	return pkg.CampaignsToJSONString(campaignsArray)
+}
+
+func Render(path string) string {
+	if path == "" {
+		return renderHomePage()
+	}
+
+	// split the path
+	parts := strings.Split(path, "/")
+	switch {
+	case len(parts) == 2 && parts[0] == "campaign":
+		return renderCampaign(parts[1])
```

Comment:
It would be a good practice (and a better customer experience) if you first check if that campaign exists before rendering something that couldn't even be an id for a campaign. (Imagine someone puts `campaign/foo`).

Reply:
![image](https://github.com/gnolang/gno/assets/111009238/eba53cca-30bd-4e03-a535-0838ce8702cc)
do you think what I did is good or not?

Reply:
Sorry, my bad. This is correct then.